File: my-folder/0105-construct-binary-tree-from-preorder-and-inorder-traversal/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        if not preorder and not inorder:
            return None
        inorder_map = defaultdict(int)
        for i in range(len(inorder)):
            inorder_map[inorder[i]] = i
        newList = TreeNode(preorder[0], None, None)
        mid = inorder_map[preorder[0]]
        newList.left = self.buildTree(preorder[1 : mid + 1], inorder[ : mid])
        newList.right = self.buildTree(preorder[mid + 1 : ], inorder[mid + 1: ])

        return newList
        




# preorder = [3,9,20,15,7]
            
# inorder = [9,3,15,20,7]
                   

File: my-folder/0105-construct-binary-tree-from-preorder-and-inorder-traversal/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    Map<Integer, Integer> mapNodes;
    int preIndex = 0;
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        mapNodes = new HashMap<>();
        for(int i = 0; i < inorder.length; i++){
            mapNodes.put(inorder[i], i);
        }
        
        return buildNodes(preorder, 0, preorder.length - 1);
    }

    private TreeNode buildNodes(int[] preorder, int left, int right){
        if(left > right){
            return null;
        }

        int root_val = preorder[preIndex++];
        TreeNode root = new TreeNode(root_val);

        int inorderIdx = mapNodes.get(root_val);

        root.left = buildNodes(preorder, left, inorderIdx - 1);
        root.right = buildNodes(preorder, inorderIdx + 1, right);

        return root;
    }

}

// pre - [3, 9, 20, 15, 7]
//        i
// in - [9, 3, 15, 20, 7]
      
